src/libpriv/rpmostree-refts.cxx

```diff
@@ -114,6 +123,29 @@ RpmTs::package_meta (const rust::Str name) const
           retval->_size = headerGetNumber (h, RPMTAG_LONGARCHIVESIZE);
           retval->_buildtime = headerGetNumber (h, RPMTAG_BUILDTIME);
           retval->_src_pkg = headerGetString (h, RPMTAG_SOURCERPM);
+
+          struct rpmtd_s nchanges_date_s;
+          _cleanup_rpmtddata_ rpmtd nchanges_date = NULL;
+          nchanges_date = &nchanges_date_s;
+          headerGet (h, RPMTAG_CHANGELOGTIME, nchanges_date, HEADERGET_MINMEM);
+          int ncnum = rpmtdCount (nchanges_date);
+          if (!ncnum)
+            ncnum = 0;
+          int frequency_recent_updates = 0;
+          GDateTime *now_dt = g_date_time_new_now_utc ();
```

Comment:
We need to be careful with looking at the current wall clock time.  It's the number one source of problems with [reproducible builds](https://reproducible-builds.org/).

Usually, instead of this we should be honoring `SOURCE_DATE_EPOCH`.  There's some already extant internal APIs for this.

Alternatively though...we could make this relative to the *most recently changed package*, which would always work even if `SOURCE_DATE_EPOCH` is not set.  But the cost would be a need to do a global pass first.

Reply:
Hmmm.  Perhaps early on in the build process we should just set `SOURCE_DATE_EPOCH` to the current time, if not set - and use that as our baseline reference always.

The huge mess with this though is [setenv isn't multithreading safe](https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475) and we definitely use threads.

So...the hack we could do here is probably just to re-exec ourselves if it's not set.

Also this relates to https://github.com/coreos/coreos-assembler/issues/3332 - probably coreos-assembler should be setting SOURCE_DATE_EPOCH if it's not set.  (But we need a reasonable fallback for non-coreos-assembler cases too, including `rpm-ostree compose image`).

Reply:
Maybe what would help the most is to keep the "heavy lifting" logic in Rust, so we'd have
`fn changelogs(self: &PackageMeta) -> Vec<u64>;`
(instead of `change_frequency`)
So what the C++ code would do is basically just build up that vector.

That would then break the dependency cycle here, because `lowest_change_time` is only known after we've gathered all the package metadata (which is what change frequency could reference).

Reply:
I'm happy to help with the C++/Rust interop bits of this if you want, it can be hairy